rtrlib/rtr_mgr.c

```diff
@@ -431,17 +525,156 @@ inline int rtr_mgr_get_spki(struct rtr_mgr_config *config,
 {
+	struct rtr_mgr_group_node *group_node;
+
+	pthread_mutex_lock(&config->mutex);
+	tommy_node *node = tommy_list_head(&config->groups);
+
 	MGR_DBG1("rtr_mgr_stop()");
-	for (unsigned int i = 0; i < config->len; i++) {
-		for (unsigned int j = 0; j < config->groups[i].sockets_len; j++)
-			rtr_stop(config->groups[i].sockets[j]);
+	while (node) {
+		group_node = node->data;
+		for (unsigned int j = 0; j < group_node->group->sockets_len;
+		     j++) {
+			rtr_stop(group_node->group->sockets[j]);
+		}
+	node = node->next;
+	}
+	pthread_mutex_unlock(&config->mutex);
+}
+
+int rtr_mgr_add_group(struct rtr_mgr_config *config,
+		      const struct rtr_mgr_group *group)
+{
+	unsigned int refresh_iv = 3600;
+	unsigned int retry_iv = 600;
+	unsigned int expire_iv = 7200;
+	int err_code = RTR_ERROR;
+
+	pthread_mutex_lock(&config->mutex);
+
+	struct rtr_mgr_group_node *gnode;
+	tommy_node *node = tommy_list_head(&config->groups);
+
+	while (node) {
+		gnode = node->data;
+		if (gnode->group->preference == group->preference) {
+			RTR_DBG("Group with preference value already exists!");
+			pthread_mutex_unlock(&config->mutex);
+			return RTR_INVALID_PARAM;
+		}
+
+		//TODO This is not pretty. It wants to get the same intervals
+		//that are being used by other groups. Maybe intervals should
+		//be store globally/per-group/per-socket?
+		if (gnode->group->sockets[0]->refresh_interval)
+			refresh_iv = gnode->group->sockets[0]->refresh_interval;
+		if (gnode->group->sockets[0]->retry_interval)
+			retry_iv = gnode->group->sockets[0]->retry_interval;
+		if (gnode->group->sockets[0]->expire_interval)
+			expire_iv = gnode->group->sockets[0]->expire_interval;
+		node = node->next;
+	}
+	struct rtr_mgr_group *new_group = malloc(sizeof(struct rtr_mgr_group));
+
+	if (!new_group) {
+		pthread_mutex_unlock(&config->mutex);
+		return RTR_ERROR;
+	}
+	memcpy(new_group, group, sizeof(struct rtr_mgr_group));
+	new_group->status = RTR_MGR_CLOSED;
+
+	for (unsigned int j = 0; j < new_group->sockets_len; j++) {
+		if (rtr_init(new_group->sockets[j], NULL, config->pfx_table,
+			     config->spki_table, refresh_iv,
+			     expire_iv, retry_iv, rtr_mgr_cb,
+			     config, new_group) != RTR_SUCCESS) {
+			err_code = RTR_INVALID_PARAM;
+			free(new_group);
+			pthread_mutex_unlock(&config->mutex);
+			return err_code;
+		}
+	}
+
+	struct rtr_mgr_group_node *new_group_node = malloc(
+					sizeof(struct rtr_mgr_group_node));
+	if (!new_group_node) {
+		pthread_mutex_unlock(&config->mutex);
+		free(new_group);
+		return RTR_ERROR;
+	}
+
+	new_group_node->group = new_group;
+	tommy_list_insert_tail(&config->groups, &new_group_node->node,
+			       new_group_node);
+	config->len++;
+
+	tommy_list_sort(&config->groups, &rtr_mgr_config_cmp);
+
+	struct rtr_mgr_group *best_group = rtr_mgr_get_first_group(config);
+
+	if (best_group->status == RTR_MGR_CLOSED) {
+		for (unsigned int j = 0; j < best_group->sockets_len; j++)
+			rtr_start(best_group->sockets[j]);
```

Comment:
What is the behavior in case one or more sockets cannot be started?
Maybe do
```
if (rtr_start(best_group->sockets[j]) != RTR_SUCCESS)
    tr_free(best_group->sockets[j]);
```
or is this not even necessary?